frame/staking-rewards/src/lib.rs

```diff
@@ -1051,58 +1116,121 @@ fn update_rewards_pool<T: Config>(
 	})
 }
 
-pub(crate) fn reward_accumulation_calculation<T: Config>(
-	reward: Reward<T::AssetId, T::Balance>,
+/// Calculates the update to the reward and unlocks the accumulated rewards from the pool account.
+pub(crate) fn do_reward_accumulation<T: Config>(
+	reward: &mut Reward<T::AssetId, T::Balance>,
+	pool_account: &T::AccountId,
 	now_seconds: u64,
-) -> Result<Reward<T::AssetId, T::Balance>, RewardAccumulationCalculationError<T>> {
-	match now_seconds.safe_sub(&reward.last_updated_timestamp) {
-		Ok(elapsed_time) => {
-			let reward_rate_period_seconds = reward.reward_rate.period.as_secs();
+) -> Result<(), RewardAccumulationCalculationError> {
+	if reward.reward_rate.amount.is_zero() {
+		return Ok(())
+	}
+
+	let elapsed_time = now_seconds
+		.safe_sub(&reward.last_updated_timestamp)
+		.map_err(|_| RewardAccumulationCalculationError::BackToTheFuture)?;
 
-			// SAFETY(benluelo): Usage of Div::div:
+	let reward_rate_period_seconds = reward.reward_rate.period.as_secs();
+
+	// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
+	// reward_rate_period_seconds is a NonZeroU64 here
+	let periods_surpassed = elapsed_time.div(reward_rate_period_seconds.get());
+
+	if periods_surpassed.is_zero() {
+		Ok(())
+	} else {
+		let total_locked_rewards: u128 =
+			T::Assets::balance_on_hold(reward.asset_id, pool_account).into();
+
+		// the maximum amount repayable given the reward rate.
+		// i.e. if total locked is 50, and the reward rate is 15, then this would be 3
+		//
+		// SAFETY(benluelo): Usage of Div::div: Integer division can only fail if rhs == 0, and
+		// reward.reward_rate.amount is known to be non-zero here as per the check at the beginning
+		// of this function
+		let maximum_releasable_periods = total_locked_rewards.div(reward.reward_rate.amount.into());
```

Comment:
Not sure why we need this, Can't we do,
```
newly_accumulated_rewards = min(periods_surpassed * reward.reward_rate.amount, total_locked_rewards)
```
To get the unlockable amount? Period related calculations seem redundant to me.

Reply:
we need to keep the accumulated amount in sync with the period, otherwise we need to track partially rewarded periods somewhere

Reply:
I see, just wondering then wouldn't it be better then to calculate and set the `last_updated_time` to be in the past to match 
`(periods_surpassed * reward.reward_rate.amount - total_locked_rewards) / reward_rate` ? Then we only need to do period related calculations in case the amount to be unlocked is larger than `total_locked_rewards`.

Reply:
It's ok to add a TODO for simplifying this and do it in a follow up PR :)